fdbserver/storageserver.actor.cpp

```diff
@@ -3476,14 +3511,19 @@ ACTOR Future<Void> replaceInterface( StorageServer* self, StorageServerInterface
 	return Void();
 }
 
-ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered )
+ACTOR Future<Void> storageServer( IKeyValueStore* persistentData, StorageServerInterface ssi, Reference<AsyncVar<ServerDBInfo>> db, std::string folder, Promise<Void> recovered)
 {
 	state StorageServer self(persistentData, db, ssi);
 	self.folder = folder;
 	self.sk = serverKeysPrefixFor( self.thisServerID ).withPrefix(systemKeys.begin);  // FFFF/serverKeys/[this server]/
 	try {
 		state double start = now();
 		TraceEvent("StorageServerRebootStart", self.thisServerID);
+		bool dispose = wait(memoryStoreRecover (persistentData, self.cx, self.thisServerID) );
+		if (dispose) {
+			TraceEvent("DisposeStorageServer", self.thisServerID);
+			throw worker_removed();
+		}
 		bool ok = wait( self.storage.restoreDurableState() );
```

Comment:
I think here you should do something like this:

```
state Future<bool> restored = self.storage.restoreDurableState();
state Future<bool> dispose = memoryStoreRecover (persistentData, self.cx, self.thisServerID);
Void _ = wait(success(restored) || success(dispose));
if (restored.isSet()) {
 // restore finished before dispose
} else {
  ASSERT(dispose.get() == true); // or change the signature of memoryStoreRecover
  // dispose database here
}
```

Reply:
one of my concern is, if this storage server can't be disposed (contains the data we need), then **canRemoveStorageServer** method will keep returning false, and we just keep retrying every second.  Will this add overhead on cluster?

Reply:
you only poll once a second - I would assume the load is very very minimal. it is a range query that fetches at most two kv-pairs